util/sock.c

```diff
@@ -297,21 +298,38 @@ sock_resolve(const char * addr)
 }
 
 /**
- * sock_resolve_one(addr):
+ * sock_resolve_one(addr, addport):
  * Return a single sock_addr structure, or NULL if there are no addresses.
  * Warn if there is more than one address, and return the first one.
+ * If ${addport} is non-zero, use sock_addr_ensure_port() to add a port number
+ * of ":0" if appropriate.
  */
 struct sock_addr *
-sock_resolve_one(const char * addr)
+sock_resolve_one(const char * addr, int addport)
 {
 	struct sock_addr ** sas;
 	struct sock_addr * sa;
 	struct sock_addr ** sa_tmp;
-
-	/* Resolve target address. */
-	if ((sas = sock_resolve(addr)) == NULL) {
-		warnp("Error resolving socket address: %s", addr);
-		goto err0;
+	char * bind_addr;
+
+	/* Resolve target address... */
+	if (addport == 0) {
```

Comment:
I would write this differently:

```
char * resolv_addr = addr;
if (addport && (resolv_addr = sock_addr_ensure_port(addr)) == NULL)) {
    ...
}
if ((sas = sock_resolve(resolv_addr)) == NULL) {
    ...
}
if (resolv_addr != addr)
    free(resolv_addr);
```

Printing errors with the original `addr` is appropriate; while it's not what we're actually resolving but it's what the user gave us and (very importantly) there shouldn't be any errors relating to `resolv_addr` which don't also apply to `addr`.

Reply:
Unfortunately your code has a "discarding const qualifier" thing, since we have `const char * addr`.

I've rewritten this commit to use `strdup()` for the non-addport case.  It's a bit ugly, but this isn't high-performance code.  And playing games with casting to & from a `(uintptr_t)` would be uglier than a `strdup`, I think.

Reply:
Ah, I forgot about that detail.  Canonical way of doing this is with an extra variable.

```
char * resolv_addr = addr;
char * resolv_addr_alloc = NULL;
if (addport && (resolv_addr = resolv_addr_alloc = sock_addr_ensure_port(addr)) == NULL)) {
    ....
}
if ((sas = sock_resolv(resolv_addr)) == NULL) {
    ....
}
free(resolv_addr_alloc);
````

Reply:
grumble, I swear that I checked this twice last night.  Huh, that's a sneaky way of doing it!

I've revised this PR accordingly, and it looks much better now.